src/net_processing.cpp

```diff
@@ -3120,8 +3120,27 @@ bool static ProcessMessage(CNode* pfrom, const std::string& strCommand, CDataStr
     }
 
     if (strCommand == NetMsgType::NOTFOUND) {
-        // We do not care about the NOTFOUND message, but logging an Unknown Command
-        // message would be undesirable as we transmit it ourselves.
+        // Remove the NOTFOUND transactions from the peer
+        LOCK(cs_main);
+        CNodeState *state = State(pfrom->GetId());
+        std::vector<CInv> vInv;
+        vRecv >> vInv;
+        if (vInv.size() <= MAX_PEER_TX_IN_FLIGHT + MAX_BLOCKS_IN_TRANSIT_PER_PEER) {
+            for (CInv &inv : vInv) {
+                if (inv.type == MSG_TX || inv.type == MSG_WITNESS_TX) {
+                    // If we receive a NOTFOUND message for a txid we requested, erase
+                    // it from our data structures for this peer.
+                    auto in_flight_it = state->m_tx_download.m_tx_in_flight.find(inv.hash);
+                    if (in_flight_it == state->m_tx_download.m_tx_in_flight.end()) {
+                        // Skip any further work if this is a spurious NOTFOUND
```

Comment:
This comment is slightly confusing in this PR, but I think the code logic is correct.  The idea is that if the tx isn't in_flight from you then there is no reason to try erasing it from `m_tx_announced`, because if its in there you are still going to deal with it at some later point via `m_tx_process_time`

Reply:
re https://github.com/bitcoin/bitcoin/pull/15834/commits/e32e08407e2781d881b9da92aa06494525ddd085#r291253714:

I think "spurious" is still fine. Why would a node announce you a tx and then follow up with a NOTFOUND without any further communication with you.

Reply:
I think the questions around why I [structured the loop this way](https://github.com/bitcoin/bitcoin/pull/15834/files/23163b759354b84c5a076e3e2ae6ae6338106035..e32e08407e2781d881b9da92aa06494525ddd085#r291901132) (and why I made a comment about skipping "any further work") is motivated by the additional code I plan to add in #15505.  See for instance: https://github.com/bitcoin/bitcoin/pull/15505/commits/f5dbb49c8922cb26d05862b7aacf161bde8ffc5a